File: 2098-subsequence-of-size-k-with-the-largest-even-sum/2098-subsequence-of-size-k-with-the-largest-even-sum.cpp
class Solution {
public:
    long long largestEvenSum(vector<int>& A, int k) { 
        long long sum = 0;
        int n = A.size();
        sort(A.begin(),A.end());
        reverse(A.begin(),A.end());
        int smallest_odd = INT_MAX;
        int smallest_even = INT_MAX;
        int i = 0;
        
        while(i < k)
        {
            sum += A[i];
            if(A[i] % 2)
                smallest_odd = min(smallest_odd,A[i]);
            else
                smallest_even = min(smallest_even,A[i]);
            i++;
        }
        if(sum % 2 == 0)
            return sum;
        
        long long res = -1;
        while(i < n)
        {
            if(A[i] % 2 and smallest_even != INT_MAX)
            {
                res = max(res,(sum - smallest_even + A[i]) );
            }
            else if(A[i] % 2 == 0 and smallest_odd != INT_MAX)
            {
                res = max(res,sum - smallest_odd + A[i] );
            }
            i++;
        }
        if(res % 2)
            return -1;
        return res;
        
    }
};

File: 2098-subsequence-of-size-k-with-the-largest-even-sum/2098-subsequence-of-size-k-with-the-largest-even-sum.java
class Solution {
    public long largestEvenSum(int[] nums, int k) {
        Arrays.sort(nums);
        int minEven = -1, minOdd = -1;
        int maxEven = -1, maxOdd = -1;
        long total = 0 , ans = -1;
        for(int i=nums.length-1;i>=nums.length-k;i--) {
            total += nums[i];    
            if(nums[i]%2==0) {
                minEven = nums[i];
            }
            else {
                minOdd = nums[i];
            }
        }
        if(total%2==0) return total;
        
        for(int i=0;i<nums.length-k;i++) {
            if(nums[i]%2==0) {
                maxEven = nums[i];
            }
            else {
                maxOdd = nums[i];
            }
        }
        
        if(minEven != -1 && maxOdd != -1) ans = Math.max(ans,total-minEven+maxOdd);
        if(minOdd != -1 && maxEven != -1) ans = Math.max(ans,total-minOdd+maxEven);
        return ans;
    }
}



